pkg/db/event_advisory_lock.go

```diff
@@ -0,0 +1,192 @@
+package db
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/openshift-online/maestro/pkg/logger"
+	"gorm.io/gorm"
+)
+
+type IsConsumerSubscribed func(consumerName string) bool
+
+// EventAdvisoryLockFactory creates advisory locks for resource events,
+// ensuring the consumer relevant to the current event is subscribed to the current gRPC broker
+// before lock creation. This should be used only when the gRPC broker is enabled.
+type EventAdvisoryLockFactory struct {
+	connection           SessionFactory
+	lockStore            *AdvisoryLockStore
+	isConsumerSubscribed IsConsumerSubscribed
+}
+
+// NewEventAdvisoryLockFactory returns a new EventAdvisoryLockFactory with an AdvisoryLockStore
+// and a function to check if the consumer relevant to the current event is subscribed to the current gRPC broker.
+func NewEventAdvisoryLockFactory(connection SessionFactory, isConsumerSubscribed IsConsumerSubscribed) *EventAdvisoryLockFactory {
```

Comment:
can we use a new controller instead of this? I think for grpc broker, we don't need a lock for spec event, when a spec event is send, the controller broadcast the event for every grpc broker, each broker only handle event with its consumers? for service status update request, we can keep the current logic

Reply:
Last week, we discussed this with @qiujian16 and @clyang82. I understand that adding a new controller (specController without lock) means each instance must handle the event and add an <event, instance> record after the instance processes the event, similar to what we did for statusController. I'm still open to this and we can discuss it further in the future.